### PR TITLE
Improve write speed of microSD card via SPI and FatFS

### PR DESCRIPTION
**Pull request by FRASTM**
This PR is adding a new function `sdhc_spi_write_multi` to increase the write operation on SD card using a SPI bus. This function is sending the CMD25 (**Write multiple blocks**) instead of looping on CMD24  (Write single blocks) like in the `sdhc_spi_write` function.
 when the nb of blocks to write is significant (more than 2).
   
There is a sample application coming from  @EminYagmahan to measure transfer speed.
This PR fixes the "write" part of the https://github.com/zephyrproject-rtos/zephyr/issues/22906.
 
Test using the ci-shield (PR #26741) with the following command:
`$ west build -b nucleo_f411re samples/subsys/fs/spi_sd/ -DSHIELD="ci_shield"`

Note that the **multiple block read** operation is already in the `sdhc_spi_read()` function which sends the SDHC_READ_MULTIPLE_BLOCK command

Signed-off-by: Francois Ramu <francois.ramu@st.com>

### CODE REVIEW

**Comment by FRASTM** [comment]
Measurements reported in the issue #22906 with this PR shows a slight improvement of the write transfer speed. 

Note that multi block transfer is for more than 2 blocks (for 1 or 2 blocks, the single mode is kept).

File: samples/subsys/fs/spi_sd/src/main.c
Comment (pabigot) [inline samples/subsys/fs/spi_sd/src/main.c:23]: This results in an unusual transfer size, neither 10^6 nor 2^20.  It could be more consistent with normal testing if the multiplier were 1024, so the total transfer size is 1 MiBy.

**Review by pabigot — DISMISSED** [review]
This contains #26319 which is blocked for requested changes to be provided.  The new material here is being reviewed, but must be rebased on master once that PR goes in.  The NAK is related to this dependency.

Some style comments, but I don't immediately see anything of functional concern.

It might be worth going back sometime and coming up with a shared sample that does file system performance the same way, rather than have each sample do it independently.

File: subsys/disk/disk_access_spi_sdhc.c
Comment (FRASTM) [reply subsys/disk/disk_access_spi_sdhc.c]: ok

File: subsys/disk/disk_access_spi_sdhc.c
Comment (FRASTM) [reply subsys/disk/disk_access_spi_sdhc.c:862]: using label "exit"

File: samples/subsys/fs/spi_sd/src/main.c
Comment (FRASTM) [reply samples/subsys/fs/spi_sd/src/main.c:23]: OK to set blocks of 1024  x 1024 bytes

**Comment by FRASTM** [comment]
update after comments from @pabigot

**Comment by pabigot** [comment]
Blocked until contained material from #26319 is merged.  I'll remove my NAK after this is rebased on master with that content.

**Comment by vaussard** [comment]
Really sorry for being slow with #26319, I will try to address comments ASAP.
But otherwise, this looks fairly orthogonal to #26319, so maybe you can even base your PR directly on master if I am too slow?

**Comment by henrikbrixandersen** [comment]
@FRASTM Please rebase on master and fix the merge conflicts.

**Comment by FRASTM** [comment]
rebase

File: samples/subsys/fs/spi_sd/README.rst
Comment (henrikbrixandersen) [inline samples/subsys/fs/spi_sd/README.rst:23]: I don't believe we have in-tree support for the `ci_shield`? Perhaps reference another board/shield?

File: subsys/disk/disk_access_spi_sdhc.c
Comment (henrikbrixandersen) [inline subsys/disk/disk_access_spi_sdhc.c:764]: This way of handling SPI SDHC CS was removed in commit c64e0ce923cde54415d203e447a1dc62894ed2b0.

File: subsys/disk/disk_access_spi_sdhc.c
Comment (henrikbrixandersen) [inline subsys/disk/disk_access_spi_sdhc.c:827]:
```suggestion
	spi_release(data->spi, &data->cfg);
```

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: subsys/disk/disk_access_spi_sdhc.c
Comment (FRASTM) [reply subsys/disk/disk_access_spi_sdhc.c:764]: replaced with 	`spi_release(data->spi, &data->cfg);`

File: subsys/disk/disk_access_spi_sdhc.c
Comment (FRASTM) [reply subsys/disk/disk_access_spi_sdhc.c]: replaced with `spi_release(data->spi, &data->cfg);`

File: samples/subsys/fs/spi_sd/README.rst
Comment (FRASTM) [reply samples/subsys/fs/spi_sd/README.rst]: this  is the purpose of PR #26741

File: boards/arm/nucleo_l4r5zi/nucleo_l4r5zi.yaml
Comment (henrikbrixandersen) [inline boards/arm/nucleo_l4r5zi/nucleo_l4r5zi.yaml:20]: Was this change intended to be included in this PR?

File: subsys/disk/disk_access_spi_sdhc.c
Comment (henrikbrixandersen) [inline subsys/disk/disk_access_spi_sdhc.c:765]: I don't believe you need to release the SPI controller here?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: subsys/disk/disk_access_spi_sdhc.c
Comment (henrikbrixandersen) [reply subsys/disk/disk_access_spi_sdhc.c]: No, `spi_release(data->spi, &data->cfg);` should only be used where the `disk_access_spi_sdhc.c` used to de-assert the CS line. I believe this call can be removed all together.

File: samples/subsys/fs/spi_sd/README.rst
Comment (henrikbrixandersen) [reply samples/subsys/fs/spi_sd/README.rst]: Yes. I understand that, but we should not reference a shield not yet in the tree.

File: boards/arm/nucleo_l4r5zi/nucleo_l4r5zi.yaml
Comment (FRASTM) [reply boards/arm/nucleo_l4r5zi/nucleo_l4r5zi.yaml]: mismatch, 
Removed

File: samples/subsys/fs/spi_sd/README.rst
Comment (FRASTM) [reply samples/subsys/fs/spi_sd/README.rst]: removing any ref to the board

File: subsys/disk/disk_access_spi_sdhc.c
Comment (henrikbrixandersen) [reply subsys/disk/disk_access_spi_sdhc.c:765]: @FRASTM ^ ?

Comment (FRASTM) [reply subsys/disk/disk_access_spi_sdhc.c]: line removed

**Comment by erwango** [comment]
@FRASTM , please rebase

File: subsys/disk/disk_access_spi_sdhc.c
Comment (henrikbrixandersen) [reply subsys/disk/disk_access_spi_sdhc.c:765]: And reintroduced?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Comment by FRASTM** [comment]
rebase

**Review by henrikbrixandersen — APPROVED** [review]

**Review by pabigot — APPROVED** [review]

**Pull request merged**
